doc/mgmt/SONiC_ConfigDB_support_Yang_default_values_and_profiles.md

```diff
@@ -0,0 +1,392 @@
+    - show running: only return user config.
+    - show running all: return user config, default value from yang model, and config from j2 template.
+  - Currently SONiC only support 'show running'
+- DB migrator has complex logic and hardcoded config data:
+  - DB migrator will be run in post startup action.
+  - DB migrator complex logic and hardcoded config for migrate from every historical version to latest version. This will keep increase and difficult to maintance.
+
+# 1 Functional Requirement
+
+## 1.1 swss-common return default value from Yang model
+
+- Return default value is optional.
+  - Application can read config without default value, also can read config with default value.
+- Backward compatibility with existing code and applications.
+
+## 1.2 swss-common return profile from profile DB
+
+- Buffer profile stored in profile tables.
+
+- Return profile is optional.
+  
+  - Application can decide read config with profile or not:
+    - When application read profile:
+      - If user overwrite the profile, user config will be return.
+      - If user not overwrite the profile, profile will be return.
+      - If user 'delete' the profile, will return nothing.
+    - When application not read profile, API will only return user config.
+      - If user overwrite the profile, user config will be return.
+      - If user not overwrite the profile, will return nothing.
+      - If user 'delete' the profile, then there also will no user config exist, will return nothing.
+
+- Backward compatibility with existing code and applications.
+  
+  - For backward compatibility when initialize buffer config from minigraph, config will be write to both config tables and profile tables.
+  - After all code migrate to use profile DB, profile will only write to profile tables.
+  - Profile support delete/revert operation:
+    - In some user scenario, user need delete a profile, and also may add config back later.
+    - For delete operation, data in profile table will not be delete, profile use PROFILE_DELETE table to handle delete/revert:
+      - When delete a profile item, profile provider will add the item key to PROFILE_DELETE table.
+      - When read profile, any key in PROFILE_DELETE will not exist in result.
+      - When user set deleted item back, the item key will be remove from PROFILE_DELETE table.
+      - For example:
+        - BUFFER_POOL table are profile:
+          
+          ```
+          "BUFFER_POOL": {
+          "ingress_lossless_pool": {
+            "size": "26531072",
+            "type": "ingress",
+            "mode": "dynamic",
+            "xoff": "6291456"
+          },
+          "egress_lossless_pool": {
+            "size": "32822528",
+            "type": "egress",
+            "mode": "static"
+          }
+          }
+          ```
+        - On mellanox devices, when set buffer mode to 'dynamic' mode, buffer manager will use swsscommon API to delete buffer pool to reclaim unused buffer resource.
```

Comment:
Just FYI, on Mellanox devices, no buffer pool or buffer profile will be removed to reclaim unused buffers.
But it's possible that a user decides to remove some buffer pool, eg.
- to migrate from double-ingress-pool mode to single-ingress-pool mode
- or to execute `config qos clear`

Reply:
Fixed, my mistake, the buffermgrd are deleting items in appldb and statedb. config DB keeps no change.